plugins/inputs/prometheus/prometheus.go

```diff
@@ -237,8 +287,151 @@ func (p *Prometheus) gatherURL(u URLAndAddress, acc telegraf.Accumulator) error
 	return nil
 }
 
+// getMesosClient returns an httpcli client configured with the available levels of
+// TLS and IAM according to flags set in the config
+func (p *Prometheus) getMesosClient() (*httpcli.Client, error) {
+	if p.mesosClient != nil {
+		return p.mesosClient, nil
+	}
+
+	uri := p.MesosAgentUrl + "/api/v1"
+	client := httpcli.New(httpcli.Endpoint(uri), httpcli.DefaultHeader("User-Agent",
+		dcosutil.GetUserAgent(p.UserAgent)))
+	cfgOpts := []httpcli.ConfigOpt{}
+	opts := []httpcli.Opt{}
+
+	var rt http.RoundTripper
+	var err error
+
+	if p.CACertificatePath != "" {
+		if rt, err = p.DCOSConfig.Transport(); err != nil {
+			return nil, fmt.Errorf("error creating transport: %s", err)
+		}
+		if p.IAMConfigPath != "" {
+			cfgOpts = append(cfgOpts, httpcli.RoundTripper(rt))
+		}
+	}
+	opts = append(opts, httpcli.Do(httpcli.With(cfgOpts...)))
+	client.With(opts...)
+
+	p.mesosClient = client
+	return client, nil
+}
+
+// getTasks requests tasks from the operator API
+func (p *Prometheus) getTasks(ctx context.Context, cli calls.Sender) (*agent.Response_GetTasks, error) {
+	resp, err := cli.Send(ctx, calls.NonStreaming(calls.GetTasks()))
+	if err != nil {
+		return nil, err
+	}
+	r, err := processResponse(resp, agent.Response_GET_TASKS)
+	if err != nil {
+		return nil, err
+	}
+
+	gs := r.GetGetTasks()
+	if gs == nil {
+		return nil, errors.New("the getTasks response from the mesos agent was empty")
+	}
+	return gs, nil
+}
+
+// processResponse reads the response from a triggered request, verifies its
+// type, and returns an agent response
+func processResponse(resp mesos.Response, t agent.Response_Type) (agent.Response, error) {
+	var r agent.Response
+	defer func() {
+		if resp != nil {
+			resp.Close()
+		}
+	}()
+	for {
+		if err := resp.Decode(&r); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return r, err
+		}
+	}
+	if r.GetType() == t {
+		return r, nil
+	} else {
+		return r, fmt.Errorf("processResponse expected type %q, got %q", t, r.GetType())
+	}
+}
+
+// getMesosTaskPrometheusURLs converts a list of tasks to a list of Prometheus
+// URLs to scrape
+func getMesosTaskPrometheusURLs(tasks *agent.Response_GetTasks) []*url.URL {
+	results := []*url.URL{}
```

Comment:
Why slice of pointers?

Reply:
I moved the processing into the parent method to avoid this.